app/main.py

```diff
@@ -1,34 +1,85 @@
+                self.decks.append(Deck(start[0], i))
+        elif start[1] == end[1]:
+            for i in range(min(start[0], end[0]), max(start[0], end[0]) + 1):
+                self.decks.append(Deck(i, start[1]))
+
+    def get_deck(self, row: int, column: int) -> Deck:
+        return next((deck for deck in self.decks if
+                     deck.row == row and deck.column == column), None)
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def fire(self, row: int, column: int) -> str:
+        if self.is_drowned:
+            return "Already Sunk!"
+        current_deck = self.get_deck(row, column)
+        if not current_deck.is_alive:
+            return "Already hit!"
+        current_deck.is_alive = False
+        if not any(deck.is_alive for deck in self.decks):
+            self.is_drowned = True
+            return "Sunk!"
+        return "Hit!"
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def __repr__(self) -> str:
+        return (f"Ship from ({self.start[0]}, {self.start[1]})"
+                f" to ({self.end[0]}, {self.end[1]})")
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: Iterable) -> None:
+        self.field = {}
+        self.ships = []
+        self.area_size = 10
+        self.area = [["   ~   " for _ in range(self.area_size)] for _ in
+                     range(self.area_size)]
+
+        for ship_coords in ships:
+            ship = Ship(ship_coords[0], ship_coords[1])
+            self.ships.append(ship)
+            for deck in ship.decks:
+                if (deck.row, deck.column) in self.field:
+                    raise ValueError(
+                        f"Invalid ship placement: "
+                        f"overlap at ({deck.row}, {deck.column})")
+                self.field[(deck.row, deck.column)] = ship
+
+            for deck in ship.decks:
+                self.area[deck.row][deck.column] = "   □   "
+
+    def print_field(self) -> None:
+        for index, ship in enumerate(self.ships):
+            print(index + 1, "Ship:", ship)
+
+    def print_area(self) -> None:
+        for row in self.area:
+            print(*row)
+
+    def fire(self, location: tuple[int, int]) -> str:
+        ship = self.field.get(location)
+        if ship:
+            return ship.fire(*location)
+        return "Miss!"
```

Comment:
Consider adding a check to ensure that the `location` is within the bounds of the `area`. If the location is out of bounds, return a more informative message like 'Out of bounds!' instead of 'Miss!'.